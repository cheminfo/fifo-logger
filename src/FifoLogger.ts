import { v4 } from '@lukeed/uuid';

import { LevelNumber, LevelWithSilent, levels } from './levels';

export type LogEntry = {
  time: number;
  level: number;
  levelLabel: string;
  uuids: string[];
  message: string;
  meta?: Record<string, any>;
  error?: Error;
};

export type FifoLoggerOptions = {
  /**
   * The maximum number of events to store.
   * @default 1000
   */
  limit?: number;
  /**
   * The minimum level of events to store.
   * Allowed values: 'fatal', 'error', 'warn', 'info', 'debug', 'trace' or 'silent'
   * @default 'info'
   *
   */
  level?: LevelWithSilent;
  /**
   * Called when a new log is added.
   */
  onChange?: (
    log: LogEntry | undefined,
    logs: LogEntry[],
    info: { depth: number },
  ) => void;
  /**
   * An object of key-value pairs to include in log lines as properties.
   */
  bindings?: Record<string, any>;
};

/**
 * A FIFO logger that stores the last events in an array.
 */
export class FifoLogger {
  private events: LogEntry[];
  private uuids: string[];
  private levelAsNumber: number;
  private limit: number;
  private bindings: Record<string, any>;
  private onChange?: (
    log: LogEntry | undefined,
    logs: LogEntry[],
    info: { depth: number },
  ) => void;
  level: LevelWithSilent;

  constructor(options: FifoLoggerOptions = {}) {
    this.uuids = [v4()];
    this.events = [];
    this.level = options.level || 'info';
    this.levelAsNumber = levels.values[this.level];
    this.limit = options.limit ?? 1000;
    this.bindings = options.bindings ?? {};
    this.onChange = options.onChange;
  }

  setLevel(level: LevelWithSilent) {
    this.level = level;
    this.levelAsNumber = levels.values[level];
  }

  setLimit(limit: number) {
    this.limit = limit;
    this.checkSize();
  }

  checkSize() {
    if (this.events.length > this.limit) {
      this.events.splice(0, this.events.length - this.limit);
    }
  }

  /**
   * Remove events from the current logger and its children.
   * @param options
   */
  clear() {
    for (let i = this.events.length - 1; i >= 0; i--) {
      if (this.events[i].uuids.includes(this.uuids[0])) {
        this.events.splice(i, 1);
      }
    }
    this.onChange?.(undefined, this.events, { depth: this.uuids.length });
  }

  getLogs(
    options: {
      minLevel?: LevelWithSilent;
      level?: LevelWithSilent;
      includeChildren?: boolean;
    } = {},
  ): LogEntry[] {
    const { level, minLevel, includeChildren } = options;
    let logs = this.events.slice();

    if (includeChildren) {
      logs = logs.filter((log) => log.uuids.includes(this.uuids[0]));
    } else {
      logs = logs.filter((log) => log.uuids[0] === this.uuids[0]);
    }

    if (level) {
      const levelNumber = Number(levels.values[level]);
      if (Number.isNaN(levelNumber)) {
        throw new Error('Invalid level');
      }
      logs = logs.filter((log) => log.level === levelNumber);
    }
    if (minLevel) {
      const levelNumber = Number(levels.values[minLevel]);
      if (Number.isNaN(levelNumber)) {
        throw new Error('Invalid level');
      }
      logs = logs.filter((log) => log.level >= levelNumber);
    }

    return logs;
  }

  /**
   * @param bindings: an object of key-value pairs to include in log lines as properties.
   * @param options: an options object that will override child logger inherited options.
   */

  child(bindings?: Record<string, any>) {
    const newFifoLogger = new FifoLogger();
    newFifoLogger.onChange = this.onChange;
    newFifoLogger.events = this.events;
    newFifoLogger.uuids = [v4(), ...this.uuids];
    newFifoLogger.level = this.level;
    newFifoLogger.bindings = { ...this.bindings, ...bindings };
    return newFifoLogger;
  }

  trace(obj: Record<string, unknown>, message: string): void;
  trace(message: string): void;
  trace(error: Error): void;
  trace(value: unknown, message?: string) {
    addEvent(this, levels.values.trace, value, message);
  }

  debug(obj: Record<string, unknown>, message: string): void;
  debug(message: string): void;
  debug(error: Error): void;
  debug(value: unknown, message?: string): void {
    addEvent(this, levels.values.debug, value, message);
  }

  info(obj: Record<string, unknown>, message: string): void;
  info(message: string): void;
  info(error: Error): void;
  info(value: unknown, message?: string): void {
    addEvent(this, levels.values.info, value, message);
  }

  warn(obj: Record<string, unknown>, message: string): void;
  warn(message: string): void;
  warn(error: Error): void;
  warn(value: unknown, message?: string): void {
    addEvent(this, levels.values.warn, value, message);
  }

  error(obj: Record<string, unknown>, message: string): void;
  error(message: string): void;
  error(error: Error): void;
  error(value: unknown, message?: string): void {
    addEvent(this, levels.values.error, value, message);
  }

  fatal(obj: Record<string, unknown>, message: string): void;
  fatal(message: string): void;
  fatal(error: Error): void;
  fatal(value: unknown, message?: string): void {
    addEvent(this, levels.values.fatal, value, message);
  }
}

function addEvent(
  logger: any,
  level: LevelNumber,
  value: unknown,
  message?: string,
) {
  if (level < logger.levelAsNumber) return;

  const event: Partial<LogEntry> = {
    level,
    levelLabel: levels.labels[level],
    time: Date.now(),
    uuids: logger.uuids,
  };
  if (value instanceof Error) {
    event.message = value.toString();
    event.error = value;
    event.meta = { ...logger.bindings };
  } else if (message && typeof value === 'object') {
    event.message = message;
    event.meta = { ...logger.bindings, ...value };
  } else if (!message && typeof value === 'string') {
    event.message = value;
    event.meta = { ...logger.bindings };
  } else {
    throw new Error('Invalid arguments');
  }

  logger.events.push(event);
  logger.checkSize();
  if (logger.onChange) {
    logger.onChange(event, logger.events, { depth: logger.uuids.length });
  }
}
