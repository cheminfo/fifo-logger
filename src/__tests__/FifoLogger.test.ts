import type { BaseLogger } from 'pino';
import { throttle } from 'throttle-debounce';
import { describe, expect, it } from 'vitest';

import { FifoLogger } from '../FifoLogger.ts';
import type { LogEntry } from '../LogEntry.ts';
import type { LogEventData } from '../events.ts';

describe('FifoLogger', () => {
  it('test pino compatibility', () => {
    /*
    interface LogFn {
        // TODO: why is this different from `obj: object` or `obj: any`?
        <T extends object>(obj: T, msg?: string, ...args: any[]): void;
        (obj: unknown, msg?: string, ...args: any[]): void;
        (msg: string, ...args: any[]): void;
    }
    */

    //@ts-expect-error should be ok
    const logger: Omit<BaseLogger, 'silent'> = new FifoLogger();

    expect(logger.info).toBeDefined();
  });

  it('simple case, default level and limit', () => {
    const logger = new FifoLogger();
    logger.trace('a trace');
    logger.debug('a debug');
    logger.info('an info');
    logger.warn('a warning');
    logger.error('an error');
    logger.fatal('a fatal error');

    const logs = logger.getLogs();

    expect(logs.map((log) => log.message)).toStrictEqual([
      'an info',
      'a warning',
      'an error',
      'a fatal error',
    ]);
    expect(removeVariableValues(logs)).toMatchSnapshot();

    const errors = logger.getLogs({ minLevel: 'error' });

    expect(errors).toHaveLength(2);

    const oneError = logger.getLogs({ level: 'error' });

    expect(oneError).toHaveLength(1);
  });

  it('simple case, limit 2', () => {
    const logger = new FifoLogger({ limit: 2 });
    logger.info('an info');
    logger.warn('a warning');
    logger.error('an error');
    logger.fatal('a fatal error');

    const logs = logger.getLogs();

    expect(logs.map((log) => log.message)).toStrictEqual([
      'an error',
      'a fatal error',
    ]);

    expect(removeVariableValues(logs)).toMatchSnapshot();

    logger.setLimit(4);
    logger.info('an info');
    logger.warn('a warning');
    logger.error('an error');
    logger.fatal('a fatal error');

    expect(logger.getLogs()).toHaveLength(4);

    logger.setLimit(2);

    expect(logger.getLogs()).toHaveLength(2);

    logger.setLevel('fatal');
    logger.info('an info');
    logger.warn('a warning');
    logger.error('an error');
    logger.fatal('a fatal error');

    expect(removeVariableValues(logger.getLogs())).toMatchSnapshot();

    const childLogger = logger.child();
    childLogger.fatal('a fatal error in a child');

    expect(logger.getLogs({ includeChildren: true })).toHaveLength(2);
    expect(
      removeVariableValues(logger.getLogs({ includeChildren: true })),
    ).toMatchSnapshot();
  });

  it('logger level', () => {
    const logger = new FifoLogger({ level: 'error' });
    logger.info('an info');
    logger.warn('a warning');
    logger.error('an error');
    logger.fatal('a fatal error');

    expect(logger.getLogs()).toHaveLength(2);

    const childLogger = logger.child();
    childLogger.warn('a warning');
    childLogger.error('an error');

    expect(childLogger.getLogs()).toHaveLength(1);

    expect(logger.getLogs({ includeChildren: true })).toHaveLength(3);
  });

  it('addEventListener', () => {
    const logger = new FifoLogger();
    const firstListener: LogEventData[] = [];
    const secondListener: string[] = [];
    const onceListener: string[] = [];
    logger.addEventListener('log', (event) => {
      firstListener.push(event.detail);
    });
    logger.addEventListener('log', (event) => {
      secondListener.push(event.detail.log.message);
    });
    logger.addEventListener(
      'log',
      (event) => {
        onceListener.push(event.detail.log.message);
      },
      { once: true },
    );
    logger.info('an info');
    logger.warn('a warning');
    logger.error('an error');
    logger.fatal('a fatal error');

    expect(firstListener.map((detail) => detail.log.message)).toStrictEqual([
      'an info',
      'a warning',
      'an error',
      'a fatal error',
    ]);
    expect(secondListener).toStrictEqual([
      'an info',
      'a warning',
      'an error',
      'a fatal error',
    ]);
    expect(onceListener).toStrictEqual(['an info']);
  });

  it('test types', () => {
    const logger = new FifoLogger();

    expect(() => {
      // @ts-expect-error should not be able to log without message
      logger.info();
    }).toThrow('Invalid arguments');
    expect(() => {
      // @ts-expect-error should not be able to log with 2 strings
      logger.info('a', 'b');
    }).toThrow('Invalid arguments');

    logger.info({}, '');

    expect(logger.getLogs()).toHaveLength(1);
  });

  it('logging message, object, error', () => {
    const logger = new FifoLogger();
    logger.info('an info');
    logger.info({ object: 'ab' }, 'an info with an object ');
    const logs = logger.getLogs();

    expect(
      logs.map((log) => ({ meta: log.meta, message: log.message })),
    ).toStrictEqual([
      { message: 'an info', meta: {} },
      { message: 'an info with an object ', meta: { object: 'ab' } },
    ]);
    expect(removeVariableValues(logs)).toMatchSnapshot();
  });

  it('child and grandchild', () => {
    const logger = new FifoLogger();
    logger.info('an info');

    // we can also get a child logger, by default it will use a random uuid
    const child = logger.child();
    child.warn('a warn in a child');

    const grandchild = child.child();
    grandchild.error('an error in a grandchild');

    child.warn('another warn in a child');

    expect(logger.getLogs()).toHaveLength(1);
    expect(logger.getLogs({ includeChildren: true })).toHaveLength(4);

    expect(child.getLogs()).toHaveLength(2);
    expect(child.getLogs({ includeChildren: true })).toHaveLength(3);
    expect(grandchild.getLogs()).toHaveLength(1);
    expect(grandchild.getLogs({ includeChildren: true })).toHaveLength(1);

    const ids = logger
      .getLogs({ includeChildren: true })
      .map((entry) => entry.id);

    expect(ids).toStrictEqual([1, 2, 3, 4]);
  });

  it('clear', () => {
    const messages: Array<string | undefined> = [];
    const depths: number[] = [];
    const logger = new FifoLogger();
    logger.addEventListener('change', (event) => {
      messages.push(event.detail.logs.at(-1)?.message);
      depths.push(event.detail.info.depth);
    });
    logger.info('an info');

    const child = logger.child();
    child.warn('a warn in a child');

    const grandchild = child.child();
    grandchild.error('an error in a grandchild');

    child.warn('another warn in a child');

    expect(logger.getLogs({ includeChildren: true })).toHaveLength(4);
    expect(child.getLogs({ includeChildren: true })).toHaveLength(3);
    expect(grandchild.getLogs({ includeChildren: true })).toHaveLength(1);

    child.clear();

    expect(logger.getLogs({ includeChildren: true })).toHaveLength(1);
    expect(child.getLogs({ includeChildren: true })).toHaveLength(0);
    expect(grandchild.getLogs({ includeChildren: true })).toHaveLength(0);

    expect(depths).toStrictEqual([1, 2, 3, 2, 2]);
    expect(messages).toStrictEqual([
      'an info',
      'a warn in a child',
      'an error in a grandchild',
      'another warn in a child',
      // After child.clear(), we are left with the first log on the root logger
      'an info',
    ]);
  });

  it('error', () => {
    const logger = new FifoLogger();
    logger.info('an info');

    logger.fatal(new Error('Fatal error'));

    const child = logger.child();
    child.fatal(new Error('fatal error in child'));

    const containsErrorObject = logger
      .getLogs({ includeChildren: true })
      .filter((log) => log.error);

    expect(containsErrorObject).toHaveLength(2);
  });

  it('child with custom properties', () => {
    const logger = new FifoLogger();
    logger.info({ a: 1 }, 'an info');

    const child = logger.child({ a: 2, b: 4 });
    child.warn('a warn in a child');
    const grandchild = child.child({ a: 3, c: 5 });
    grandchild.error('an error in a grandchild');

    const metas = logger
      .getLogs({ includeChildren: true })
      .map((log) => log.meta);

    expect(metas).toStrictEqual([
      { a: 1 },
      { a: 2, b: 4 },
      { a: 3, b: 4, c: 5 },
    ]);
  });

  it('bindings', () => {
    // it is possible to have some default parameters for all logs and this can be overwriten by the child

    const logger = new FifoLogger({ bindings: { namespace: 'base' } });
    logger.warn('a warning with object');
    logger.warn({ c: 'Hello' }, 'a warning');

    const loggerChild = logger.child({ namespace: 'child', a: 1 });
    loggerChild.info('from loggerChild');
    const loggerSubChild = loggerChild.child({ namespace: 'subchild', b: 2 });
    loggerSubChild.info('from loggerSubChild');

    expect(
      logger
        .getLogs({ includeChildren: true })
        .map((log) => ({ message: log.message, meta: log.meta })),
    ).toMatchInlineSnapshot(`
      [
        {
          "message": "a warning with object",
          "meta": {
            "namespace": "base",
          },
        },
        {
          "message": "a warning",
          "meta": {
            "c": "Hello",
            "namespace": "base",
          },
        },
        {
          "message": "from loggerChild",
          "meta": {
            "a": 1,
            "namespace": "child",
          },
        },
        {
          "message": "from loggerSubChild",
          "meta": {
            "a": 1,
            "b": 2,
            "namespace": "subchild",
          },
        },
      ]
    `);
  });

  it('onchange', () => {
    const results: Array<string | number | undefined> = [];
    const logger = new FifoLogger();
    logger.addEventListener('change', (event) => {
      results.push(event.detail.logs.at(-1)?.message, event.detail.logs.length);
    });
    logger.info('first info');
    logger.info('second info');

    expect(results).toStrictEqual(['first info', 1, 'second info', 2]);

    const childLogger = logger.child();
    childLogger.info('info in child');

    expect(results).toStrictEqual([
      'first info',
      1,
      'second info',
      2,
      'info in child',
      3,
    ]);
  });

  it('onchange takes only depth 1', () => {
    const results: Array<string | number> = [];
    const logger = new FifoLogger();
    logger.addEventListener('change', (event) => {
      const log = event.detail.logs.at(-1);
      const info = event.detail.info;
      if (log && info.depth === 1) {
        results.push(log.message, event.detail.logs.length);
      }
    });
    logger.info('first info');
    logger.info('second info');

    expect(results).toStrictEqual(['first info', 1, 'second info', 2]);

    const childLogger = logger.child();
    childLogger.info('info in child');

    expect(results).toStrictEqual(['first info', 1, 'second info', 2]);
  });

  it('onchange with throttle', () => {
    const results: Array<number | string> = [];
    const throttleFunc = throttle(100, (log, logs) => {
      results.push(log.message, logs.length);
    });
    const logger = new FifoLogger();
    logger.addEventListener('change', (event) => {
      const log = event.detail.logs.at(-1);
      throttleFunc(log, event.detail.logs);
    });
    logger.info('first info');
    logger.info('second info');
    const start = Date.now();
    while (Date.now() - start < 120);
    logger.info('an info after 120ms');

    expect(results).toStrictEqual(['first info', 1, 'an info after 120ms', 3]);
  });
});

function removeVariableValues(logs: LogEntry[]) {
  return logs.map((log) => ({
    ...log,
    time: 42,
    uuids: log.uuids.length,
    error: log.error ? { message: 'Message', name: 'Name' } : undefined,
  }));
}
