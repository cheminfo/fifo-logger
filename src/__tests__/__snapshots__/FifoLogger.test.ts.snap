// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FifoLogger logging message, object, error 1`] = `
[
  {
    "error": undefined,
    "id": 1,
    "level": 30,
    "levelLabel": "info",
    "message": "an info",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 2,
    "level": 30,
    "levelLabel": "info",
    "message": "an info with an object ",
    "meta": {
      "object": "ab",
    },
    "time": 42,
    "uuids": 1,
  },
]
`;

exports[`FifoLogger simple case, default level and limit 1`] = `
[
  {
    "error": undefined,
    "id": 1,
    "level": 30,
    "levelLabel": "info",
    "message": "an info",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 2,
    "level": 40,
    "levelLabel": "warn",
    "message": "a warning",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 3,
    "level": 50,
    "levelLabel": "error",
    "message": "an error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 4,
    "level": 60,
    "levelLabel": "fatal",
    "message": "a fatal error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
]
`;

exports[`FifoLogger simple case, limit 2 1`] = `
[
  {
    "error": undefined,
    "id": 3,
    "level": 50,
    "levelLabel": "error",
    "message": "an error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 4,
    "level": 60,
    "levelLabel": "fatal",
    "message": "a fatal error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
]
`;

exports[`FifoLogger simple case, limit 2 2`] = `
[
  {
    "error": undefined,
    "id": 8,
    "level": 60,
    "levelLabel": "fatal",
    "message": "a fatal error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 9,
    "level": 60,
    "levelLabel": "fatal",
    "message": "a fatal error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
]
`;

exports[`FifoLogger simple case, limit 2 3`] = `
[
  {
    "error": undefined,
    "id": 9,
    "level": 60,
    "levelLabel": "fatal",
    "message": "a fatal error",
    "meta": {},
    "time": 42,
    "uuids": 1,
  },
  {
    "error": undefined,
    "id": 10,
    "level": 60,
    "levelLabel": "fatal",
    "message": "a fatal error in a child",
    "meta": {},
    "time": 42,
    "uuids": 2,
  },
]
`;
